diff --git a/include/asm-generic/vmlinux.lds.h b/include/asm-generic/vmlinux.lds.h
index 0d5b186ab..36aafc54c 100644
--- a/include/asm-generic/vmlinux.lds.h
+++ b/include/asm-generic/vmlinux.lds.h
@@ -141,6 +141,7 @@ defined(CONFIG_AUTOFDO_CLANG) || defined(CONFIG_PROPELLER_CLANG)
 	*(__rt_sched_class)			\
 	*(__fair_sched_class)			\
 	*(__ext_sched_class)			\
+	*(__hvf_sched_class)			\
 	*(__idle_sched_class)			\
 	__sched_class_lowest = .;
 
diff --git a/include/uapi/linux/sched.h b/include/uapi/linux/sched.h
index 359a14cc7..1ce5d5b1d 100644
--- a/include/uapi/linux/sched.h
+++ b/include/uapi/linux/sched.h
@@ -119,6 +119,7 @@ struct clone_args {
 #define SCHED_IDLE		5
 #define SCHED_DEADLINE		6
 #define SCHED_EXT		7
+#define SCHED_HVF		8
 
 /* Can be ORed in to make sure the process is reverted back to SCHED_NORMAL on fork */
 #define SCHED_RESET_ON_FORK     0x40000000
diff --git a/kernel/sched/build_policy.c b/kernel/sched/build_policy.c
index fae1f5c92..b294ad1a3 100644
--- a/kernel/sched/build_policy.c
+++ b/kernel/sched/build_policy.c
@@ -51,6 +51,8 @@
 
 #include "rt.c"
 
+#include "hvf.c"
+
 #ifdef CONFIG_SMP
 # include "cpudeadline.c"
 # include "pelt.c"
diff --git a/kernel/sched/hvf.c b/kernel/sched/hvf.c
new file mode 100644
index 000000000..e5ceeed8f
--- /dev/null
+++ b/kernel/sched/hvf.c
@@ -0,0 +1,13 @@
+#include <linux/sched.h>
+#include <asm-generic/errno.h>
+#include "sched.h"
+
+const struct sched_class hvf_sched_class;
+
+
+
+
+
+DEFINE_SCHED_CLASS(hvf)={
+
+};
